app/src/main/java/org/thoughtcrime/securesms/net/ChunkedDataFetcher.java

```diff
@@ -0,0 +1,405 @@
+          compositeController.cancel();
+          return;
+        }
+
+        Optional<Long> contentLength = parseLengthFromContentRange(contentRange);
+
+        if (!contentLength.isPresent()) {
+          Log.w(TAG, "Unable to parse length from Content-Range.");
+          callback.onFailure(new IOException("Unable to get parse length from Content-Range."));
+          compositeController.cancel();
+          return;
+        }
+
+        if (chunkSize >= contentLength.get()) {
+          try {
+            callback.onSuccess(response.body().byteStream());
+          } catch (IOException e) {
+            callback.onFailure(e);
+            compositeController.cancel();
+          }
+        } else {
+          InputStream stream = ContentLengthInputStream.obtain(response.body().byteStream(), chunkSize);
+          fetchChunks(url, contentLength.get(), Optional.of(new Pair<>(stream, chunkSize)), compositeController, callback);
+        }
+      }
+    });
+
+    return compositeController;
+  }
+
+  private void fetchChunks(@NonNull String url,
+                           long contentLength,
+                           Optional<Pair<InputStream, Long>> firstChunk,
+                           CompositeRequestController compositeController,
+                           Callback callback)
+  {
+    List<ByteRange> requestPattern;
+    try {
+      if (firstChunk.isPresent()) {
+        requestPattern = Stream.of(getRequestPattern(contentLength - firstChunk.get().second()))
+                               .map(b -> new ByteRange(b.start + firstChunk.get().second(),
+                                                       b.end   + firstChunk.get().second(),
+                                                       b.ignoreFirst))
+                               .toList();
+      } else {
+        requestPattern = getRequestPattern(contentLength);
+      }
+    } catch (IOException e) {
+      callback.onFailure(e);
+      compositeController.cancel();
+      return;
+    }
+
+    SignalExecutors.UNBOUNDED.execute(() -> {
+      List<CallRequestController> controllers = Stream.of(requestPattern).map(range -> makeChunkRequest(client, url, range)).toList();
+      List<InputStream>           streams     = new ArrayList<>(controllers.size() + (firstChunk.isPresent() ? 1 : 0));
+
+      if (firstChunk.isPresent()) {
+        streams.add(firstChunk.get().first());
+      }
+
+      Stream.of(controllers).forEach(compositeController::addController);
+
+      for (CallRequestController controller : controllers) {
+        Optional<InputStream> stream = controller.getStream();
+
+        if (!stream.isPresent()) {
+          Log.w(TAG, "Stream was canceled.");
+          callback.onFailure(new IOException("Failure"));
+          compositeController.cancel();
+          return;
+        }
+
+        streams.add(stream.get());
+      }
+
+      try {
+        callback.onSuccess(new InputStreamList(streams));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *InputStreamList* constructor in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).